### ISSUE TITLE
Unable to connect to MongoDB

### ISSUE THREAD

**Issue by benstrum** [aws/studio-lab-examples#51]
I am trying to connect to MongoDB Atlas.  Are certain ports blocked?

from pymongo import MongoClient
MONGO_URL = "xxx"
client = MongoClient(MONGO_URL)
client.db.find({})

Results in ServerSelectionTimeoutError on port 27017

**Comment by EmilyWebber** [comment]
Hi, we're taking a look at this right now.

**Comment by swagulkarni** [comment]
@benstrum - Couple of questions:

1. Are you trying to connect to a MongoDB installed elsewhere from the notebook?
2. Which notebook are you trying to test?

**Comment by benstrum** [comment]
Hi, I had created my own notebook where I was going to pull a dataset out
of one of my existing mongo servers.  To reproduce the issue, you can
easily create a free mongo db at their site (
https://www.mongodb.com/cloud/atlas/register) and then try to connect.  My
guess is that port 27017 is blocked on aws' end.

-Ben

**Comment by EmilyWebber** [comment]
How are you managing your connection to your Mongo instance? Can you validate that the connection works with the same code outside of SMSL, for example from your laptop?

**Comment by benstrum** [comment]
Yes I'm able to connect locally and from Google colabs.  Have you tried
making a connection?  As i mentioned, you can create a free account in
mongo atlas   Also, can you verify that the port is open?

**Comment by darrenkraker** [comment]
This issue also exists for outbound traffic on port 3306.  We have a notebook that connects to a mysql RDS instance that is open to the world (academic use with fake data) works great from EC2 or other db tools but the same code when run from a Jupyter Lab notebook fails to connect.  Running curl -v telnet://<dbendpoint>.cluster-xxxxxx.us-west-2.rds.amazonaws.com:3306 times out when run from a terminal session on the Jupyter Lab notebook but not my local instance.  Are there any outbound network restrictions?  Port 443 seems to work but not 22 or 3306?

**Comment by darrenkraker** [comment]
So I found a work around if you're willing to use a reverse TCP proxy using NGINX.  It looks like the notebooks will support outbound traffic on port 80.  I setup a reverse proxy with the following config in nginx.conf
stream {
    server {
        listen     80;
        #TCP traffic will be forwarded to the specified server
        proxy_pass .cluster-xxxxx.us-west-2.rds.amazonaws.com:3306;
    }
}
Then updated my notebook to connect as follows:
try:
    mydb = mysql.connector.connect(
      host="ec2-xx-xx-xx-xx.us-west-2.compute.amazonaws.com", # URL or hostname of the database in Oregon  
      user="dbuser",
      password="<PASSWORD>",
      port="80",
      database='XXX')
      
It would be great if the Jupyter Studio Lab environment would allow outbound 3306, 27017, etc . . 
Maybe if we are nice to @EmilyWebber she can help us fix it ;)

**Comment by EmilyWebber** [comment]
@darrenkraker It never hurts ... haha just kidding! We'll try to get some clarity on which ports are open / closed for SMSL straight away.

**Comment by benofben** [comment]
We're having the same issue with it blocking Neo4j ports.  Apparently 80, 443 and 53 are open.  Keep in mind this isn't incoming traffic it's block but outgoing, which is pretty strict....

I've tried to configure to use those other ports but AWS Linux seems to have things running on them already, so I'm getting bind failures.

**Comment by jasonbub** [comment]
Hi @EmilyWebber , 

I just wanted to chime in here.  It is now close to 1 year since @benofben wrote the excellent note above that describes the issue in detail.  

I just spent about 1 hour trying to figure out why a connection that works from other AWS systems and Google Colab doesn't work in Studio Lab.  I then luckily stumbled on this thread which saved me from another several hours of searching for a solution.  I now know that SageMaker Studio Lab only supports a limited set of ports externally, but what is being done about this?  

Any chance we will get access other ports or is Studio Lab not a viable solution for those of us that need to access ports other than 80, 443 and 53?

Thanks in advance for your time!

**Comment by MicheleMonclova** [comment]
Hi, these ports were purposely closed for security reasons. 
My best recommendation if you need these ports now is to migrate to SageMaker Studio.  It requires an AWS account. There is a free tier that you can take advantage of.
